designs/SpanMarshallers.md

```diff
@@ -0,0 +1,247 @@
+
+```csharp
+[GenericCollectionMarshaller]
+public struct GenericCollectionMarshallerImpl<T, U, V,...>
+{
+    // these constructors are required if marshalling from managed to native is supported.
+    public GenericCollectionMarshallerImpl(GenericCollection<T, U, V, ...> collection, int nativeSizeOfElements);
+    public GenericCollectionMarshallerImpl(GenericCollection<T, U, V, ...> collection, Span<byte> stackSpace, int nativeSizeOfElements); // optional
+    
+    public const int StackBufferSize = /* */; // required if the span-based constructor is supplied.
+
+    // These method is required if marshalling from managed to native is supported.
+    public TCollectionElement GetManagedValueAtIndex(int i);
+    // This method is required if marshalling from native to managed is supported.
+    public void SetManagedValueAtIndex(int i, TCollectionElement value);
+
+    // The getter is required if marshalling from managed to native is supported.
+    // The setter is required if marshalling from native to managed is supported.
+    public int Count { get; set; }
+
+    // The requirements on the Value property are the same as when used with `NativeTypeMarshallingAttribute`.
+    // The property is required with the generic collection marshalling.
+    public TNative Value { get; set; }
+
+    public ref byte GetOffsetForNativeValueAtIndex(int index);
+}
+```
+
+The constructors now require an additional `int` parameter specifying the native size of the collection elements, represented as `TCollectionElement` above. As the elements may be marshalled to types with different native sizes than managed, this enables the author of the generic collection marshaller to not need to know how to marshal the elements of the collection, just the collection structure itself.
+
+The `GetManagedValueAtIndex` method and `Count` getter are used in the process of marshalling from managed to native. The generated code will iterate through `Count` elements (retrieved through `GetManagedValueAtIndex`) and assign their marshalled result to the address represented by `GetOffsetForNativeValueAtIndex` called with the same index. Then either the `Value` property getter will be called or the marshaller's `GetPinnableReference` method will be called, depending on if pinning is supported in the current scenario.
+
+The `SetManagedValueAtIndex` method and the `Count` setter are used in the process of marshalling from native to managed. The `Count` property will be set to the number of elements that the native collection contains, and the `Value` property will be assigned the result value from native code. Then the stub will iterate through the native collection `Count` times, calling `GetOffsetForNativeValueAtIndex` to get the offset of the native value and calling `SetManagedValueAtIndex` to set the unmarshalled managed value at that index.
```

Comment:
I would be very interested in an example on matrices here - particularly higher order ones.